OAuth: Authomatic -> Flask-Dance

drryanjames opened the pull request:
These changes remove Authomatic and use Flask-Dance in its place to authenticate users as they access the portal. These changes include:

- Several tests and an updated backdoor
- Removal of authomatic references
- Refactoring of oauth logic to fit into flask-dance's framework

At a high level, Flask-Dance uses blueprints to authenticate with providers and returns control when auth succeeds or fails. If auth succeeds, these changes use the bearer token to get info from the provider (i.e. Facebook or Google). FlaskDanceProvider is an abstract base class that's used to get this info. The MockFlaskDanceProvider mocks this call by returning data passed through the test backdoor.

Once the user's info is received we perform the same logic as before: update the database and log the user in. I reordered some of the database logic to align with flask-dance's tutorials, but it's pretty much the same as before.

These changes can be tested at [ryan-dev.cirg.washington.edu](http://ryan-dev.cirg.washington.edu).

I'm not completely sure about this, but to upgrade a production environment I think you'll want to:
- Change the names of the facebook and google settings in application.cfg
- `export OAUTHLIB_RELAX_TOKEN_SCOPE=1` ([more info](https://flask-dance.readthedocs.io/en/latest/quickstarts/google.html?highlight=OAUTHLIB_RELAX_TOKEN_SCOPE))
- `flask db upgrade`

There may be other things to check as well. If there is, please let me know. I'd like to learn.

drryanjames commented on instance/application.cfg.default:
```diff
@@ -6,7 +6,8 @@ SERVER_NAME = '<the name and port number of the server (e.g.: 'myapp.dev:5000')>
-# FB_CONSUMER_SECRET = '<App Secret from Facebook>'
-# GOOGLE_CONSUMER_KEY = '<App ID from Google>'
-# GOOGLE_CONSUMER_SECRET = '<App Secret from Google>'
+# FACEBOOK_OAUTH_CLIENT_ID = '<App ID from Facebook>'
```
@ivan-c just calling out that I changed the names of these settings.

drryanjames commented on portal/views/auth.py:
```diff
@@ -43,13 +45,430 @@
+)
+
+
+class FlaskDanceProvider:
```
Any guidance on how to organize the code in this file would be much appreciated. I just kinda threw stuff in here because I wasn't sure where else to put it. I was considering the below structure but wanted to get some feedback first.

```
/portal
    /views
        /auth
            /consumer <-- Flask-Dance related files
                facebook.py
                google.py
                consumerAuth.py <-- General functions/apis
            /provider <- Flask-OAuthlib (existing code)
                *.py <- intervention/app related stuff
    /models
        */**/*.py <- similar structure here where applicable
```

drryanjames commented:
@pbugni @ivan-c, would you mind doing a quick code review?

drryanjames commented:
Just found [this](https://flask-dance.readthedocs.io/en/latest/testing.html)--a way to run pytests without the need for an auth backdoor. I think the backdoor we had/have is good enough, but the method described in the link would make our source and test code a bit cleaner and remove the need for our auth backdoor which is a potential security risk.

Any votes for whether I should spend time adding this in? It would probably take a day.

drryanjames commented on portal/views/auth.py:
```diff
@@ -43,13 +45,430 @@
+
+
+class MockFlaskDanceProvider(FlaskDanceProvider):
+    """creates user info from test data to validate auth logic
```
This class is used by the backdoor API later in this file. If I move this class to the test module I'll need to import it from the test module into this file, which would would make our production code dependent on the test module which would be a weird dependency.

pbugni replied:
Fair enough!

pep8speaks commented:
Hello @drryanjames! Thanks for updating the PR.











 - In the file [`portal/models/flaskdanceprovider.py`](https://github.com/uwcirg/true_nth_usa_portal/blob/4e41ebdcd72ef81ce5e680c1e093f8095a2c9217/portal/models/flaskdanceprovider.py), following are the PEP8 issues :

> [Line 165:1](https://github.com/uwcirg/true_nth_usa_portal/blob/4e41ebdcd72ef81ce5e680c1e093f8095a2c9217/portal/models/flaskdanceprovider.py#L165): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

drryanjames commented:
@pbugni thanks for the review comments! I added documentation and removed references to the old config. This help me spot a route that I didn't test--when a user uninstalls our app the deauthorized callback is called, which was using the old FB_CONSUMER_SECRET config key.

drryanjames commented:
@pbugni thanks for signing off. I don't have access to squash and merge this myself. Would you mind doing so for me (or inviting me to the repo)? Here is what I'm thinking for the merge comments:

* Update 3rd party sign in: Replace Authomatic with Add Flask-Dance
* Create simple and extendable architecture to login users and handle errors
* Update Google and Facebook OAuth app config keys
* Add end-to-end tests
* Update Documentation

drryanjames commented on docs/source/provider_authentication.rst:
```diff
@@ -0,0 +1,78 @@
+-  Set the ``Authorized redirect URIs`` to exactly match the location of ``<scheme>://<hostname>/login/google/``
+-  Enable the ``Google+ API``
+
+managep.py
```
That's not how you spell manage? :P

drryanjames commented on docs/source/provider_authentication.rst:
```diff
@@ -0,0 +1,78 @@
+
+managep.py
+~~~~~~~~~~
+In a **non-production** environment add the following to ``manage.py``:
```
yes! Thanks. I'll update

drryanjames commented on portal/config/config.py:
```diff
@@ -130,11 +130,6 @@ class BaseConfig(object):
     STAFF_BULK_DATA_ACCESS = True
     PATIENT_LIST_ADDL_FIELDS = []  # 'status', 'reports'
 
-    FB_CONSUMER_KEY = os.environ.get('FB_CONSUMER_KEY')
```
Will do. I didn't realize they were still used after I deleted authomatic

drryanjames commented on portal/models/auth.py:
```diff
@@ -27,6 +29,9 @@ class AuthProvider(db.Model):
     user = db.relationship('User')
 
+    __table_args__ = (UniqueConstraint('provider', 'provider_id',
+                                       name='auth_providers_by_provider'),)
```
I prefer the single indentation too but was having trouble passing pep8squeaks and the auto indent tool I used did this. Will update.

drryanjames replied:
Actually, I'm not sure why I double indented this (and a few other places) because I was clearly passing pep8sqeaks with a single indent in other places. Nice catch.

drryanjames commented on portal/migrations/versions/a031e26dc1bd_.py:
```diff
@@ -0,0 +1,44 @@
+
+    op.add_column('auth_providers',
+                  sa.Column('token',
+                            sqlalchemy_utils.types.json.JSONType(),
```
Yes, I believe so. The [flask dance base class](https://github.com/singingwolfboy/flask-dance/blob/master/flask_dance/consumer/backend/sqla.py#L40) enforces JSON

drryanjames commented on portal/views/auth.py:
```diff
@@ -30,26 +31,286 @@
+    """
+    if request.args.get('next'):
+        session['next'] = request.args.get('next')
+        validate_origin(session['next'])
```
Probably. I tried to change as little as possible from the old logic, but reordering this makes sense to me too. I'll make the update.

drryanjames commented on portal/views/auth.py:
```diff
@@ -30,26 +31,286 @@
+    error_description=None,
+    error_uri=None
+):
+    """handles outh errors
```
I need a spell checker 😢

drryanjames commented on tests/__init__.py:
```diff
@@ -171,17 +185,33 @@ def promote_user(self, user=None, role_name=None):
+
+        # Use the key value pairs in oauth_info dict
+        # as query params
+        oauth_url = 'test/oauth?' + urllib.parse.urlencode(oauth_info)
```
ooooo, nice! Will do
